### PR TITLE
Fixed issue #15037: Data protection statement texts not saved if text for checkbox is too long

### PR DESCRIPTION
**Pull request by gabrieljenik**
Fixed issue #15037: Data protection statement texts not save if text for checkbox is too long.

Added maxlength = 192 in Data policy settings view for data sec label (editDataSecurityLocalSettings_view.php).
Also added a control in the controller (database.php): if there is an error while saving the language settings, the error is shown and the survey is not saved.

### CODE REVIEW

**Comment by Shnoulle** [comment]
https://github.com/LimeSurvey/LimeSurvey/pull/1435/files#diff-05b8c46923ca0091db7ee406d881eb5bR34 192 character only … 😧

**Comment by gabrieljenik** [comment]
Yes, that comes from the DB

**Comment by Shnoulle** [comment]
All DB ?

**Comment by gabrieljenik** [comment]
https://github.com/LimeSurvey/LimeSurvey/blob/071204150cb4a5affbc1532c9ca1936238b88467/application/helpers/update/updatedb_helper.php#L2220

**Comment by Shnoulle** [comment]
Beurk … ;)

**Comment by Shnoulle** [comment]
@gabrieljenik : why not adding it in rules ?

Then : more clean when import or use API.

Without your fix : it's done only for HTML GUI part, must be fixed too (not throw SQL error) when import and

**Comment by gabrieljenik** [comment]
@Shnoulle The rule was already there.
` array('surveyls_policy_notice_label', 'length', 'min' => 0, 'max'=>192),`

Part of the issue was that when the text was too loong the saving failure was not reported back with an error message.

**Comment by Shnoulle** [comment]
Thank you :)

File: application/controllers/admin/database.php [hunk 1]
```diff
@@ -814,6 +814,8 @@ private function actionUpdateSurveyLocaleSettings($iSurveyID)
         $languagelist = $oSurvey->additionalLanguages;
         $languagelist[] = $oSurvey->language;
 
+        $saveSurvey = true;
```
Comment (olleharstedt) [inline application/controllers/admin/database.php:817]: Please do not introduce needless state into the logic. This is spaghetti code. Instead, throw an exception if survey can't be saved.

File: application/controllers/admin/database.php [hunk 2]
```diff
@@ -881,129 +883,137 @@ private function actionUpdateSurveyLocaleSettings($iSurveyID)
                     if (count($data) > 0) {
                         $oSurveyLanguageSetting = SurveyLanguageSetting::model()->findByPk(array('surveyls_survey_id'=>$iSurveyID, 'surveyls_language'=>$langname));
                         $oSurveyLanguageSetting->setAttributes($data);
-                        $oSurveyLanguageSetting->save(); // save the change to database
+                        // Save the change to database
+                        if (!$oSurveyLanguageSetting->save()) { 
+                            $saveSurvey = false;
+                            Yii::app()->setFlashMessage(Chtml::errorSummary($oSurveyLanguageSetting,Chtml::tag("p",array('class'=>'strong'),gT("Survey could not be updated, please fix the following error:"))), "error");
```
Comment (olleharstedt) [inline application/controllers/admin/database.php:889]: Throw exception here, and add flash message in the `catch` block instead.

Comment (olleharstedt) [reply application/controllers/admin/database.php:889]: Or just throw an exception and let Yii error handler show it.

Comment (olleharstedt) [reply application/controllers/admin/database.php:889]: OK, I'm getting some understanding of your fix. database.php is _not_ an easy file to work with, with so big methods. I pushed a fix where I split it into two methods instead, and added a try-catch block around those methods.

Comment (gabrieljenik) [reply application/controllers/admin/database.php:889]: Nice! Will take into account the advice. Agree: Sometime code reorgs are needed.

Comment (olleharstedt) [reply application/controllers/admin/database.php:889]: Yeah, not an easy call as consultant. In-house I can do "bolder" changes. But you can always ask us.

Comment (gabrieljenik) [reply application/controllers/admin/database.php:889]: Yes, when it come to fixes,I try to keep the scope tight

Comment (olleharstedt) [reply application/controllers/admin/database.php:889]: That's exactly right, as small fix as possible (with any given interpretation of "possible" ^^ ), to make cherry-picking between branches easier.

**Pull request merged**